### PR TITLE
fix(android): prevent internal 'current keyboard' desynchronization state

### LINKED ISSUE #6703: bug(android): getKeyboardInfo with invalid index: -1
Sentry Issue: [KEYMAN-ANDROID-75](https://sentry.io/organizations/keyman/issues/2720587980/?referrer=github_integration)

```
getKeyboardInfo with invalid index: -1

    at dalvik.system.VMStack.getThreadStackTrace(VMStack.java)
    at java.lang.Thread.getStackTrace(Thread.java:1736)
    at com.tavultesoft.kmea.util.KMLog.LogError(KMLog.java:39)
    at com.tavultesoft.kmea.data.KeyboardController.getKeyboardInfo(KeyboardController.java:170)
    at com.tavultesoft.kmea.KMManager.getCurrentKeyboardInfo(KMManager.java:1747)
    at com.keyman.android.SystemKeyboard.onStartInput(SystemKeyboard.java:142)
    at android.inputmethodservice.InputMethodService.doStartInput(InputMethodService.java:2433)
    at android.inputmethodservice.InputMethodService$InputMethodImpl.startInput(InputMethodService.java:652)
    at android.inputmethodservice.InputMethodService$InputMethodImpl.dispatchStartInputWithToken(InputMethodService.java:688)
    at android.inputmethodservice.IInputMethodWrapper.executeMessage(IInputMethodWrapper.java:199)
    at com.android.internal.os.HandlerCaller$MyHandler.handleMessage(HandlerCaller.java:44)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:257)
    at android.app.ActivityThread.main(ActivityThread.java:8192)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:626)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1015)
```

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #6703.
~Supersedes /~ replaces #6704.

Well... further investigation into the base issue while attempting to better verify my thoughts at the time... has shown that my conclusions were incorrect.  I'll leave this up for now in case further investigations can transform this into something useful after all, but there's a strong chance this may end up a "closed" PR.

~After a significant investigation into related code, it appears that we've actually had a subtle race condition in system keyboard code.  It is possible for the privacy check added by #2255 to be run _before_ the system keyboard's host page has loaded, as the host page's `keyboardLoaded` function is an async callback from its WebView.  (We can't guarantee that this will complete before the system calls the SystemKeyboard's `onStartInput` function.)~

~Correspondingly... it is theoretically possible for the predictive-text banner to be _active_ when the race condition triggers... which could result in a password being displayed as a default suggestion (but otherwise not stored or transmitted off-device) by improperly-active predictive text if the user's active keyboard has a matching lexical model.  So... we should make sure to get this right and ensure that the privacy-check runs.~

~It is notable that a number of conditions, as noted above, are required for users to experience a significant impact.  However... that impact is severe and potentially costly to the user.~

My main investigation notes regarding this may be found in the base issue - #6703.

### Other Notes

A small shout-out to `git blame` for pointing out the relevance of #2255 when I stopped to think why the code block was even there in the first place.  We'd have totally missed this angle otherwise.

# User Testing

Unfortunately, as this is caused from a race condition... there's probably no way to be fully sure that it addresses the issue.

* TEST_BASIC: Run through a basic usage test of Keyman for Android, both in-app and system keyboards, verifying that you can switch keyboards, lock and unlock your device, switch between Keyman and other input methods, and install and use Keyman keyboards without errors.
* TEST_PASSWORD_LOCK_SCREEN: using `sil_euro_latin` with predictive text enabled, ensure that password entry is secure when using Keyman to input a password on your lock screen
    * For reference, you can set up Keyman as the default system keyboard via the "Getting Started" menu in-app.
    * Make sure that `sil_euro_latin` displays a predictive text banner when loaded within the app.
    * Ensure your device's lock screen is configured to require a password input via keyboard for unlocking.
    * With all of the above in place:
        1.  Lock your device.
        2. Wait at least 5 seconds before continuing.
        3. "Wake" your device, starting the unlock process.  The Keyman keyboard should be displayed; if not, check all point above and retry.
        4. If the predictive text banner is active, _**FAIL**_ this test.
        5. Fully unlock your device.
    * Run the sequence above at least ten times.
        * Again, any attempt's step-`d` failure will _**FAIL**_ this test.

### CODE REVIEW

**Comment by mcdurdin** [comment]
> there's probably no way to be fully sure that it addresses the issue

You could use a delay timer in a custom build to force a specific order of operations for test purposes?

**Comment by jahorton** [comment]
I may need to investigate a little more deeply to ensure that the banner actually _is_ disabled should the suspected race condition trigger; I don't think we can do the JS calls properly before the page fully loads, now that I think about it in more detail.  @mcdurdin made a good suggestion above that should also prove useful for such an investigation.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 1]
```diff
@@ -1994,7 +1994,25 @@ public static void resetContext(KeyboardType kbType) {
...
+    // the user's last selected keyboard.
+    //
+    // For documentation purposes:  "before" can happen for the system keyboard; there's no guarantee
+    // that the host page will finish its async load before the `onStartInput` method is called.
+    if(key == null) {
+      SharedPreferences prefs = context.getSharedPreferences(context.getString(R.string.kma_prefs_name), Context.MODE_PRIVATE);
+      int index = prefs.getInt(KMManager.KMKey_UserKeyboardIndex, 0);
+      if (index < 0) {
+        index = 0;
+      }
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2011]: This shifts logic from `pageLoaded` to a shared function which could be called anywhere. I am not very comfortable with that. What about the situation where there are no keyboards loaded when this is called? That's probably not normal for Keyman but it is possible with 3rd party apps that embed KMEA. In this case, index 0 would be an invalid index and cause crashes elsewhere.

**Review by mcdurdin — CHANGES_REQUESTED** [review]

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 1 above)
Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2011]: > What about the situation where there are no keyboards loaded when this is called?

This case is completely unchanged by this PR.  Note how the `if(key == null)` part doesn't actually reference any keyboards, loaded or not -- just the cached index value from "shared preferences".

The part that actually references keyboards was already in the function.

> In this case, index 0 would be an invalid index and cause crashes elsewhere.

That's the very case we're addressing with this PR.  It was already in this function.

Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2011]: > What about the situation where there are no keyboards loaded when this is called?

So... issue #6703?  The issue this PR is addressing?  That's the old code in this function.  

Note how the the section - the `if(key == null)` part - doesn't actually reference any keyboards, loaded or not -- just the cached index value from "shared preferences".  It's there _explicitly_ to handle this case.

Following from that, this 'reduces' down to the same concern as your other comment.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 2]
```diff
@@ -1994,7 +1994,25 @@ public static void resetContext(KeyboardType kbType) {
...
+      return index;
+    } else {
+      return KeyboardController.getInstance().getKeyboardIndex(key);
```
[the remark this answers is not in the document]
Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2014]: Eh, to be fair, if this case happens, we _really_ need there to be an error log about it.  That indicates a state of internal inconsistency.

I would like to point out that in the app's current state... its callers _do_ handle that case.  They just log that the inconsistency occurred, which is what led us to the issue in the first place.

Not that we shouldn't double-check that there's not more we can do without casing these inconsistencies to become silent errors.  But, we shouldn't make this case become a silent error.  Should that state occur, something's still pretty "off" internally.

[the remark this answers is not in the document]
Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2014]: Sure, logging is good, I want to see that. But I don't see the need to crash the app if we can recover. In this PR, not all its callers do handle that case. For example, LanguageSettingsActivity.java / onClick will crash:

https://github.com/keymanapp/keyman/blob/411c79fd53beba8160fa6b287bf44a827097cf6a/android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/LanguageSettingsActivity.java#L91

Also:

https://github.com/keymanapp/keyman/blob/411c79fd53beba8160fa6b287bf44a827097cf6a/android/KMEA/app/src/main/java/com/tavultesoft/kmea/ModelPickerActivity.java#L212

Also, if we return an index of `0`, where there is no keyboard installed, that gets passed into (eventually) `KeyboardController.getKeyboardInfo()`. We won't get that logged into Sentry, as it's only logged to console (`Log.w()`):

https://github.com/keymanapp/keyman/blob/411c79fd53beba8160fa6b287bf44a827097cf6a/android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/KeyboardController.java#L165-L183

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 1 above)
Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2011]: I don't think you understand what I am saying. If there are no keyboards installed here, we should not be setting an index of `0`. We should be staying with an index of `-1`, which is a valid index, equivalent to `null`.

Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2011]: Okay, that definitely makes your underlying question _way_ clearer than before.  Right, I didn't think that much about the 3rd party use case.

Admittedly... I'd been thinking of extracting it to a separate function originally, and then I ran into the nomenclature problem - "what do I name this?"  Turns out the name I chose... already had an implementation.  And it seemed to slot in as a very natural fit, to boot.

Sounds like I should move the `0`-check back to its original location, then.

The line before the block seen above:

```
int index = prefs.getInt(KMManager.KMKey_UserKeyboardIndex, 0);
```

So... should we have this default to `-1` as well, then?  Though... is there really a scenario outside the initial load of the app where there wouldn't be some previously-cached non-negative keyboard index?  Do we actually allow 3rd party apps to do _that_?

Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2011]: Yeah, this is tricky: if no default has ever been selected, it probably makes sense to return 0 -- assuming there is a keyboard in that slot. 

A slightly more nuanced, more robust, and probably better solution:

1. If no keyboards installed, return -1.
2. Get the pref value, default 0 (reasonable to default to the first keyboard if never previously checked).
3. If value < 0, make it 0 (so we always try for a valid keyboard).
4. Then, verify that it's a valid keyboard index.
5. If not a valid keyboard index, set to 0 (and update the pref??).

I think that means that this function will always return "good" data which seems like a reasonable design to me 😂: Good data means that if the index >= 0 it is guaranteed at this time to be pointing to an installed keyboard, and -1 always means no keyboards installed.

**Comment by jahorton** [comment]
After further deep diving into related code and the base issue... it appears my concerns re: password exposure were a false alarm.  Which is a good thing overall!

... which then means that we still don't understand the underlying cause(s) that led to the original issue.  (Obviously not  good, but it's at least less "bad" than the good news is "good".)

**Pull request closed without merging**